### DISCUSSION TITLE
Generate code in rust buildscript instead of through CLI?

### DISCUSSION THREAD

**Discussion by CJBuchel** [fzyzcjy/flutter_rust_bridge#2159] [Q&A] [answered]
I have a project that uses a rust server and a flutter frontend. Currently for communication I've been using a strange mix of serde/schemars to generate json schema outputs, and then flutter side use an npm package to turn it into dart classes. And all that is tied to the build-script in an infrastructure crate in my local project, so it generates everything whenever I do a build. Bar the flutter side where I have to run the npm command.
 
Recently I've had a need for logic that I need both rust side and flutter, and I've also been meaning to switch my strange build script converter to a nicer one which doesn't need json as a medium, or npm to convert for that matter XD. But I would like to keep the buildscript style of converting all my code like my json schemars.

I've seen quite a bit of documentation on the `flutter_rust_bridge_codegen` CLI tool being used with input and output options. But I was curious if it was possible to add the `flutter_rust_bridge_codegen` to my buildscript as a build dependency and run the generate function from it instead of me manually having to install and run the `flutter_rust_bridge_codegen generate`. It would make life a bit easier so I didn't have to worry about adding another dependency to the list of things I need to install before developing on a new machine. And, it runs in the script each time there is a change to my infrastructure crate which is nice.

TL;DR, i did see that there was a function from the `flutter_rust_bridge_codegen` crate just named generate, `lib_flutter_rust_bridge_codegen::codegen::generate(config, metadata)` but I'm unsure if it's intended to be used in the way I'm trying to force it to be used. Is it safe to try and use it for generating everything I need as if it were just the commands? And for the config, how would I set it up to run with specific rust input and dart output locations?

**Comment by fzyzcjy** [comment, marked as answer]
Yes, it's possible to be in build.rs: https://cjycode.com/flutter_rust_bridge/guides/misc-features/build-rs

However, currently when in this mode, the macros will not be expanded. This may or may not be a problem for you, depending on whether your code has macros. Also feel free to PR if you need the macros and need build.rs!

**Reply by CJBuchel** [reply]
Ah, somehow I missed that build.rs documentation while skimming it over. I'll give it a try!
I've been doing a few tests in the meantime to see if the generate works in my build.rs and it seems to be freezing and locking up my build when modying the Config struct directly. But I'll give the yaml a go with the build.rs shown in the example.

**Reply by CJBuchel** [reply]
So I still seem to get the same issue. My build is locking up.
Using the example yaml it complains about using the old styling and to use the new syntax, so I've chucked down
```yml
# See `pure_dart` example for comments on the configs
rust_input: crate::api
rust_root: tms-infra/tms-infra-logic/
dart_output: tms-client/lib/generated
local: true
```
But in doing so, it seems to freeze the build. (For reference the tms-infra/tms-infra-logic/ project is a direct clone of the simple example provided, so `crate::api` should be correct)

Is there anything else I'm missing? verbose logging doesn't provide any extra info that I can see, just seems to freeze.
Happy to open an issue for this

**Reply by fzyzcjy** [reply]
Hmm, IIRC we do not support macro is because if we call `cargo expand` to parse macros, it will freeze. Not sure whether it is this freeze though.

So firslty, can you run the https://github.com/fzyzcjy/flutter_rust_bridge/tree/master/frb_example/dart_build_rs example?

Secondly, if you are using the master version, try to bisect or debug which line it stucks (e.g. by adding a lot of `log::info`, or by using a debugger).

**Reply by CJBuchel** [reply]
I didn't get round to debugging it to figure out the line it was having an issue on, but I did try and replicate the example project as much as possible.

And it seems to be fixed now, it wasn't happy having the rust root as a separate crate for some reason. So when I bundled my `tms-infra` and my `tms-infra-logic` crate into one with both the `flutter_rust_bridge_codegen` and the `flutter_rust_bridge` dependencies it started to work fine. Unsure if this was the actual issue or not. Or possibly just some kind of namespace issue, I noticed that there wasn't a check on `crate::api` and it would freeze regardless of what I put there as the rust_input, so it's possible I put the wrong namespace in and for some reason it freezed up.

I'm not particularly sure. But it does work now, so thanks

**Reply by fzyzcjy** [reply]
You are welcome!
